pkg/instance/instance.go

```diff
@@ -406,122 +413,205 @@ func (s *Server) spdkInstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceSe
+			return fmt.Errorf("continuously receiving errors for %v times, stopping watching SPDK replicas", maxMonitorRetryCount)
+		}
+
 		select {
 		case <-ctx.Done():
-			logrus.Info("Stopped watching SPDK replica")
-			return nil
+			logrus.Info("Stopped watching SPDK replicas")
+			return ctx.Err()
 		default:
 			_, err := notifier.Recv()
 			if err != nil {
+				status, ok := grpcstatus.FromError(err)
+				if ok && status.Code() == grpccodes.Canceled {
+					logrus.WithError(err).Warn("SPDK replica watch is canceled")
+					return err
+				}
 				logrus.WithError(err).Error("Failed to receive next item in SPDK replica watch")
+				time.Sleep(pollInterval)
+				failureCount++
 			} else {
-				if err := srv.Send(&empty.Empty{}); err != nil {
-					return errors.Wrap(err, "failed to send instance response")
-				}
+				notifyChan <- struct{}{}
 			}
 		}
 	}
 }
 
-func (s *Server) watchSPDKEngine(ctx context.Context, req *emptypb.Empty, srv rpc.InstanceService_InstanceWatchServer) error {
-	c, err := spdkclient.NewSPDKClient(s.spdkServiceAddress)
-	if err != nil {
-		return grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to create SPDK client").Error())
-	}
-	defer c.Close()
+func (s *Server) watchSPDKEngine(ctx context.Context, req *emptypb.Empty, client *spdkclient.SPDKClient, notifyChan chan struct{}) error {
+	logrus.Info("Start watching SPDK engines")
 
-	notifier, err := c.EngineWatch(context.Background())
+	notifier, err := client.EngineWatch(context.Background())
 	if err != nil {
 		return errors.Wrap(err, "failed to create SPDK engine watch notifier")
 	}
 
+	failureCount := 0
 	for {
+		if failureCount >= maxMonitorRetryCount {
+			logrus.Errorf("Continuously receiving errors for %v times, stopping watching SPDK engines", maxMonitorRetryCount)
+			return fmt.Errorf("continuously receiving errors for %v times, stopping watching SPDK engines", maxMonitorRetryCount)
+		}
+
 		select {
 		case <-ctx.Done():
-			logrus.Info("Stopped watching SPDK engine")
-			return nil
+			logrus.Info("Stopped watching SPDK engines")
+			return ctx.Err()
 		default:
 			_, err := notifier.Recv()
 			if err != nil {
+				status, ok := grpcstatus.FromError(err)
+				if ok && status.Code() == grpccodes.Canceled {
+					logrus.WithError(err).Warn("SPDK engine watch is canceled")
+					return err
+				}
 				logrus.WithError(err).Error("Failed to receive next item in SPDK engine watch")
+				time.Sleep(pollInterval)
+				failureCount++
 			} else {
-				if err := srv.Send(&empty.Empty{}); err != nil {
-					return errors.Wrap(err, "failed to send instance response")
-				}
+				notifyChan <- struct{}{}
 			}
 		}
 	}
 }
 
-func (s *Server) watchProcess(ctx context.Context, req *emptypb.Empty, srv rpc.InstanceService_InstanceWatchServer) error {
+func (s *Server) watchProcess(ctx context.Context, req *emptypb.Empty, client *client.ProcessManagerClient, notifyChan chan struct{}) error {
 	logrus.Info("Start watching processes")
```

Comment:
nit: These watch methods look quite similar. Could consolidate them later.

Reply:
NP! Thank you.